Add game loop

bean00 opened the pull request:
- Checks if the board is full (*not if a player won)
- **Question**: what do you think about making `play-round` a private function?
    - It is public right now, but only being used in `play-game`
    - I would mainly keep it public, because the tests are more detailed and thorough (the tests for `play-game` just make sure certain functions are called

Commit:
Add game loop

- Checks if the board is full (*not if a player won)

jdesrosiers commented on src/clojure_tic_tac_toe/game_handler.clj:
```diff
@@ -0,0 +1,44 @@
+        updated-board (board/add-move board move player)
+        game-is-finished (is-game-finished? updated-board)]
+    (io/display-board updated-board)
+    (zipmap [:board :player :finished?]
```
`zipmap` shouldn't be necessary.  You can just use a map literal.

bean00 replied:
changed

jdesrosiers replied:
Don't just tell us it's done.  Push the commit that fixes it.

jdesrosiers reviewed (changes requested):
Think about how you can split your `game_handler` namespace.  Can you define your game logic functions without IO?  Or, maybe with arguments allowing you to inject IO?  Then you can create a second namespace that uses those functions to build a console version of the game.  Theoretically, you should be able to do the same to build a GUI verison of the game.

If you can make that split, I think it will be a lot more clear which things should be public and which should be private.

bean00 commented:
@jdesrosiers, I refactored out the `console_ui` from the `game_handler`, rebased the commit, and force pushed it. Please review and accept the changes if complete (*I will make the other changes for #5 on #5's branch).

tbaik commented:
@bean00, I read over Jason's comment above and looked at your most recent code. To me, it seems like you misunderstood what Jason was trying to get at.

For me, your console_ui still has the issue that Jason is pointing out. We want to decouple the game logic from the io logic and be able to swap out the io really easily. It seems like you coupled the two inside of console_ui.

I think this is the sort of thing that Jason was expecting. Correct me if I'm wrong @jdesrosiers :

`(play-game console-io board players)` and `(play-game web-io board players)`

jdesrosiers commented:
@tbaik Jon implemented this the way I intended.  In fact, we paired on this refactor.  Your example is one of the suggestions I gave for him to think about, but I don't think it is ideal.

Different types of UI can vary in so many different ways that I don't think it is possible to just inject a UI into the game loop.  For example, if we a have TTT REST API with a web front-end, there is no explicit game loop.  The loop is requests and responses between the client and server.  There is no way that can fit into the explicit loop in `play-game`.  This tells me that this is a UI detail rather than a game logic detail.

Instead, I pushed Jon in the direction of having a namespace of game logic functions that can be used when constructing a UI of any shape.  I think he has mostly achieved that goal.
